gitlab: 12.1.6 -> 12.3.5, bundler: 1.17.2 -> 1.17.3

talyz opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
- Update GitLab to 12.3.2
- Update update.py to cope with the new upstream repository structure
- Refactor gitlab-shell to use buildGoPackage and bundlerEnv for
  dependencies
- Make update.py able to update gitlab-shell dependencies
- Various fixes necessary for update to work

This would replace  #69325

###### Things done
Manually built and tested both ce and ee on fresh nixops VirtualBox VMs. I would like to test it more thoroughly, but it seems to be working well in my quick tests. 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @flokli @globin @bgamari @krav

talyz committed:
gitlab: 12.1.6 -> 12.3.4

- Update GitLab to 12.3.4

- Update update.py to cope with the new upstream repository structure

- Refactor gitlab-shell to use buildGoPackage and bundlerEnv for
  dependencies

- Refactor gitlab-workhorse to use buildGoPackage for dependencies

- Make update.py able to update gitlab-shell and gitlab-workhorse
  dependencies

- Various fixes necessary for update to work

bgamari commented on pkgs/applications/version-management/gitlab/gitlab-shell/default.nix:
```diff
@@ -1,41 +1,61 @@
+    goDeps = ./deps.nix;
 
+    postConfigure = ''
+      rm -r "$NIX_BUILD_TOP/go/src/gitlab.com/gitlab-org/gitaly/vendor/"
```
This deserves a comment. I suspect this is what I was missing in #68908.

talyz replied:
It did indeed solve that issue. :) Good call, I'll add comments.

bgamari commented on pkgs/applications/version-management/gitlab/gitlab-workhorse/default.nix:
```diff
@@ -1,21 +1,27 @@
   makeFlags = [ "PREFIX=$(out)" "VERSION=${version}" "GOCACHE=$(TMPDIR)/go-cache" ];
 
+  postConfigure = ''
+    rm -r "$NIX_BUILD_TOP/go/src/gitlab.com/gitlab-org/labkit/vendor"
```
This also needs a comment.

bgamari reviewed (commented):
This looks great. I'll try testing it tonight.

bgamari commented:
I suspect that `gitlab-shell` still isn't quite right. Trying to push to a repository via SSH appears to fail with:
```
$ git push
/nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/lib/gitlab_shell.rb:270:in `repo_path=': Repository path not provided. Please make sure you're using GitLab v8.10 or later. (ArgumentError)
	from /nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/lib/gitlab_shell.rb:108:in `verify_access'
	from /nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/lib/gitlab_shell.rb:42:in `block in exec'
	from /nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/lib/gitlab_metrics.rb:50:in `measure'
	from /nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/lib/gitlab_shell.rb:42:in `exec'
	from /nix/store/wdr7r9vzshv61x0mw00i1gbnggy23vpi-gitlab-shell-7.1.2/bin/gitlab-shell:20:in `<main>'
fatal: Could not read from remote repository.
```

talyz commented:
@bgamari Ah, I'll see if I get the same result..

bgamari commented:
Actually, looking at my GitLab user's authorized keys file it refers to `gitlab-shell-7.2.1` (my test box was quite out-of-date). I suspect `authorized_keys` just needs to be [rebuilt](https://docs.gitlab.com/ce/administration/raketasks/maintenance.html#rebuild-authorized_keys-file).

bgamari commented:
Indeed, false alarm. Simply running
```
sudo -u git gitlab-rake gitlab:shell:setup
```
resolved the issue. Things appear to be just peachy.

talyz commented:
Yeah, I wasn't able to reproduce it on my end, so that makes sense. :)

flokli commented on pkgs/applications/version-management/gitlab/data.json:
```diff
@@ -1,32 +1,32 @@
     "owner": "gitlab-org",
-    "repo": "gitlab-ce",
-    "rev": "v12.1.6",
+    "repo": "gitlab-foss",
```
Do we really want to keep separate attrsets? It seems like `gitlab-foss` is just a poorly-synced version of the `gitlab` repo these days...

talyz replied:
Maybe not? It doesn't seem to be too poorly synced to me, but from what I've gathered, the only difference between them is that the `gitlab` repo has an `ee` directory which is removed from the tree when it's synced to `gitlab-foss`...

flokli replied:
Well, at least the git repo history is a bit less informative.

Obviously, Nix doesn't care much. However, after reading up on https://about.gitlab.com/2019/02/21/merging-ce-and-ee-codebases/, it should be safe to just remove the `/ee` folder if we're not building the enterprise edition, so I'd advocate for having a conditional step in `postUnpack` depending on the `gitlabEnterprise` flag, and hardcode everything else to https://gitlab.com/gitlab-org/gitlab.

This would greatly simplify data.nix and the updater.

talyz replied:
Yeah, that states pretty clearly that all we have to do is get rid of the `ee` folder, so I'll look into it. :)

alyssais commented:
I don't know anything about GitLab, but the Bundler update looked good
to me, so I've pushed that to master as 96a1dbac972.

talyz committed:
gitlab-shell: Split patch into ruby and go parts

Split the remove-hardcoded-locations patch into two separate patches,
one for the ruby package and one for the go package. This is clearer
and results in fewer rebuilds.

talyz commented:
I just pushed some minor changes:
- Added comments as suggested by @bgamari in https://github.com/NixOS/nixpkgs/pull/70216#discussion_r330315651
- Indentation fixes and removal of commented-out code in `gitlab-shell`
- Split the patch in `gitlab-shell` into go and ruby parts
- Rebased on master

bgamari commented:
Whoops, I accidentally commented on the old PR.

In short, currently the NixOS systemd unit does not clear `/run/gitlab/config/initializers` which can cause issues with this upgrade.

Also, there are some very strange [migration issues](https://gitlab.com/gitlab-org/gitlab/issues/33262) which I had to work around.

flokli commented:
Can we bump this to 12.3.3? There has been another security release…

petabyteboy commented:
I have tested the gitlab commit and it works nicely :+1:

talyz commented:
@flokli ..and now they've released 12.3.4! I'll update to it once the deb has been built.

@bgamari Hm, yeah, that should definitely be purged on startup. I feel like it fits better into #68721, which already tries to address similar migration issues concerning the state directory. That PR also needs a bit of a push. :)

bgamari commented:
For what it's worth I have tested this and it works well.

talyz commented:
I just pushed some updates:
- Updated to 12.3.4
- Added support for updating the go dependencies of `gitlab-workhorse` in `update.py`

flokli commented:
@talyz can you address https://github.com/NixOS/nixpkgs/pull/70216#discussion_r330747432 too?

talyz commented:
@flokli Oh, sorry, I completely missed that. Will do.

flokli commented:
Thanks!

petabyteboy commented:
Here goes 12.3.5..

talyz committed:
gitlab: Build frontend assets from source

Instead of extracting prebuilt assets from the debian build, build
them from the source. This should give faster package updates and
reduces the amount of data needed to be downloaded by more than 500MB.

flokli commented:
@talyz poke about the repository updates ;-)

talyz commented:
@flokli Yeah, I'm on it! :)

talyz committed:
gitlab: Refactor for new repo structure

GitLab recently restructured their repos; whereas previously they had
one gitlab-ce and one gitlab-ee repo, they're now one and the
same. All proprietary components are put into the ee subdirectory -
removing it gives us the foss / community version of GitLab. For more
info, see
https://about.gitlab.com/2019/02/21/merging-ce-and-ee-codebases/

This gives us the opportunity to simplify things quite a bit, since we
don't have to keep track of two separate versions of either the base
data or rubyEnv.

talyz committed:
gitlab: 12.3.4 -> 12.3.5

talyz committed:
gitlab: Add myself to list of maintainers

lheckemann commented:
This isn't building, because apparently the hash is wrong:
```
trying https://gitlab.com/api/v4/projects/gitlab-org%2Fgitlab-workhorse/repository/archive.tar.gz?sha=v8.10.0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  196k  100  196k    0     0   912k      0 --:--:-- --:--:-- --:--:--  912k
unpacking source archive /private/tmp/nix-build-source.drv-0/archive.tar.gz?sha=v8.10.0
hash mismatch in fixed-output derivation '/nix/store/yr8w0isia86rg96wl4f2ikbg4h8ahw2h-source':
  wanted: sha256:11cfhh48dga5ghfcijb68gbx0nfr5bs3vvp2j1gam9ac37fpvk0x
  got:    sha256:07hv2kl2w1gw9613mvz844vy2dc4my4d8n9b36j25hyaw564zj82
```
(from hydra: https://hydra.nixos.org/build/102785558/nixlog/18 )

Could you have a look at that?

talyz commented:
@lheckemann Hm, strange. I'll look into it right away.

talyz commented:
@lheckemann Hm, it looks like the hash is different on macOS, for some reason? I can't reproduce this error on NixOS, at least. I also don't have any mac with just the right version of macOS at hand, so I can't verify that this is what happens. If this is the case, I think the reasonable thing to do is to limit `gitlab-workhorse`, and possibly the others, to `linux`; this is the restriction put on `gitlab` already and I don't see why you would want to host the other components on a different platform... That said, if `fetchFromGitlab` sometimes returns different hashes for different platforms, this should be investigated.

zimbatm commented:
@talyz try changing the hash to something different and re-run the build. If your /nix/store already has a fixed-output derivation with the given hash it won't try and re-download the dependency and you won't see the issue.

talyz commented:
@zimbatm Yeah, I know - I ran `nix-collect-garbage` to get rid of any build in my store, but it still built without any problem. Then, just to be sure, I changed the hash to a hash that is probably wrong, which caused a hash mismatch that suggests that the right hash is the one that was already specified in the file. I've also tried it on one additional NixOS machine and I managed to find a mac to test it on - it builds fine there too. Additionally, the `aarch64-linux` build on hydra works - https://hydra.nixos.org/build/102785354, and it uses the same source. So yeah, I don't really know how to take this any further - someone with access to hydra probably has to look into it.

petabyteboy commented:
> @talyz try changing the hash to something different and re-run the build. If your /nix/store already has a fixed-output derivation with the given hash it won't try and re-download the dependency and you won't see the issue.

I can not reproduce the issue either even after cleaning the store.

talyz commented:
@lheckemann @zimbatm Well, I guess this solved itself, or someone poked around in hydra :)

zimbatm commented:
hehe, oh well :)
